Replace app.js with webpack-dev-server

scottbell opened the pull request:
<!--- Note: Please open the PR in draft form until you are ready for active review. -->
Closes #5031  and #4922 and #5476

### Describe your changes:

- Uses [webpack-dev-server](https://webpack.js.org/configuration/dev-server/) instead of our custom `app.js` express application.
- Cleans up our use of NODE_ENV variables
- Updates our testing configurations to avoid using HMR when collecting code coverage
- Removes the need for cross_env
- Removes our Memory size override for Node when running the unit tests as node 14 has increased the default to our previously specified size

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [ ] Tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [x] Changes appear to address issue?
* [x] Changes appear not to be breaking changes?
* [x] Appropriate unit tests included?
* [x] Code style and in-line documentation are appropriate?
* [x] Commit messages meet standards?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

unlikelyzero commented:
@scottbell looks like all the tests pass and code coverage continues to generate (semi-)correctly.

1. Can you do one more pass to remove all app.js dependencies
2. Can you clean up the `webpack.common.js`, `karma.conf.js` and `webpack.prod.js` scripts to double check we've removed all vestigial bits.
3. Can you document our use of webpack env flags: production, test, coverage, etc. somewhere?

akhenry commented:
@scottbell  Wow, ok, this was surprisingly painless :) Thanks for doing this, it will remove a lot of ambiguity about the role of app.js, which was only ever intended as a development server, and removes a bunch of dependencies. It will also avoid a lot of spurious security alerts.

Agree with @unlikelyzero on removing the dependencies. Once that's done I'll approve and merge.

Thanks again!

scottbell commented on e2e/tests/visual/controlledClock.visual.spec.js:
```diff
@@ -22,7 +22,7 @@
 /*
 Collection of Visual Tests set to run in a default context. The tests within this suite
-are only meant to run against openmct's app.js started by `npm run start` within the
+are only meant to run against openmct started by `npm run start` within the
```
typo here, should just be `npm start`

scottbell commented on e2e/tests/framework/baseFixtures.e2e.spec.js:
```diff
@@ -23,7 +23,7 @@
 This test suite is dedicated to testing our use of the playwright framework as it
 relates to how we've extended it (i.e. ./e2e/baseFixtures.js) and assumptions made in our dev environment
-(app.js and ./e2e/webpack-dev-middleware.js)
+("npm start" and ./e2e/webpack-dev-middleware.js)
```
`npm start`

scottbell commented on e2e/tests/visual/default.visual.spec.js:
```diff
@@ -22,7 +22,7 @@
 /*
 Collection of Visual Tests set to run in a default context. The tests within this suite
-are only meant to run against openmct's app.js started by `npm run start` within the
+are only meant to run against openmct started by `npm run start` within the
```
typo again

unlikelyzero commented:
On second pass, I'm not seeing the e2e code coverage being generated.

Also this is being thrown from the e2e tests

[webpack-dev-server] Content not from webpack is served from '/root/project/dist

scottbell commented:
> @scottbell looks like all the tests pass and code coverage continues to generate (semi-)correctly.
> 
>     1. Can you do one more pass to remove all app.js dependencies? `minimist` `require` `express` were documented, but there may be more
> 
>     2. Can you clean up the `webpack.common.js`, `karma.conf.js` and `webpack.prod.js` scripts to double check we've removed all vestigial bits.
> 
>     3. Can you document our use of webpack env flags: production, test, coverage, etc. somewhere?

I've added some documentation on how we use different webpack configurations. I took another pass, and I couldn't find anymore libraries that `app.js` was using. Note `minimist` is still being used by `gendocs.js`.

scottbell commented:
> [webpack-dev-server] Content not from webpack is served from '/root/project/dist

That's expected and normal.

>I'm not seeing the e2e code coverage being generated
I'm not sure how this was done - did it involve `app.js`?

scottbell commented:
> @scottbell Wow, ok, this was surprisingly painless :) Thanks for doing this, it will remove a lot of ambiguity about the role of app.js, which was only ever intended as a development server, and removes a bunch of dependencies. It will also avoid a lot of spurious security alerts.
> 
> Agree with @unlikelyzero on removing the dependencies. Once that's done I'll approve and merge.
> 
> Thanks again!

Done!

scottbell commented:
@unlikelyzero 
> I'm not seeing the e2e code coverage being generated

How does one know the e2e code coverage isn't being generated? The e2e code coverage report looks similar to me as other PRs:
<img width="279" alt="Screen Shot 2022-09-22 at 1 25 24 PM" src="https://user-images.githubusercontent.com/9853862/191734642-9dfe5172-0653-4659-9e2b-978350cdb109.png">

scottbell commented:
@unlikelyzero Ah, there was this in `app.js`:
```
if (__DEV__) {
    webpackConfig = require('./webpack.dev');
    webpackConfig.plugins.push(new webpack.HotModuleReplacementPlugin());
    webpackConfig.entry.openmct = [
        'webpack-hot-middleware/client?reload=true',
        webpackConfig.entry.openmct
    ];
    webpackConfig.plugins.push(new WatchRunPlugin());
} else {
    webpackConfig = require('./webpack.coverage');
}
```

Note the `webpack.coverage` business at the end. We could effect this using an npm script like so:
`npx webpack serve --config ./webpack.coverage.js`

scottbell commented:
@unlikelyzero I've added a new script called `start:coverage` that starts the `webpack-dev-server` using the `webpack.coverage.js` configuration instead. I've also modified the various playwright configs to use this instead of `npm start`. Let me know what you think.

unlikelyzero commented:
Ok now it's all coming back to me: the reason that this task was going to be difficult is because we need to capture the different ways this project is built and bundled. App.js gave us the ability to keep things like hmr configurable at build and runtime.

Here's a 5am braindump that we need to document as a part of this PR.

We were using env flags that are interpreted at buildtime *and* runtime: development, test, and production flags.

Our app.js and webpack config has 4 or 5 modes of operation that we need to document outside of the configs. 

1. Production
- does this need to be the default when using npm start?
- what sourcemap config is necessary?
- production env flag

2. Development
-  does this need to be the default when using npm start?
- hmr enabled
- development env flag

3. Unit tests
- which webpack config is necessary?
- which sourcemap config? Does this conflict with the karma config?
- which runtime configs are necessary?
- hmr enabled
- I think we should always generate coverage. This used to be configured at runtime at some point 

4. e2e suite
- which webpack config is necessary?
- hmr disabled
- Always generate code coverage

5. Perf tests
- not yet implemented
- hmr disabled
- production mode?

scottbell commented:
@unlikelyzero I'll add my thoughts on your braindump below:

> App.js gave us the ability to keep things like hmr configurable at build and runtime.

You can configure HMR for webpack-dev-server [pretty easily](https://webpack.js.org/guides/hot-module-replacement). It's enabled by default. I think it's only a runtime thing?

> 
>     Production
> 
>     does this need to be the default when using npm start?
>     what sourcemap config should be used?
>     production env flag

I think this should not be the default. `npm start` is meant for development purposes only (per the ticket). Regarding sourcemaps, [webpack recommends](https://webpack.js.org/configuration/devtool/)) they should not be included. Do we need the environment flags? From what I can tell, we're using to configure CI stuff, but not much else. I think the CI stuff could also be managed through explicit script calls.

>     Development
> 
>     does this need to be the default when using npm start?
>     hmr enabled
>     development env flag

Yes, this should be the default per the ticket.

>     Unit tests
> 
>     which webpack config is necessary?
>     which sourcemap config? Does this conflict with the karma config?
>     which runtime configs are necessary?
>     hmr enabled
>     I think we should always generate coverage. This used to be configured at runtime at some point
>     test or coverage flag

For unit tests, it's using `webpack.coverage.js`. The sourcemap are still pretty garbled up due to this unfortunately, and make it tough for developers to debug. We do get (mostly) the right line numbers now though.

>     e2e suite
> 
>     which webpack config is necessary?
>     hmr disabled
>     Always generate code coverage
>     -test or coverage flag

As the e2e tests need test coverage, they're using `webpack.coverage.js`. Why do we want `hmr` disabled - does it matter?


>     Perf tests
> 
>     not yet implemented
>     hmr disabled
>     production mode?

I'm not sure about performance testing as I haven't been involved. 😀

unlikelyzero commented:
@scottbell I mean that we should document this behavior outside of the webpack, e2e, and karma files and in some place for others to reference. 

https://github.com/nasa/openmct#building-and-running-open-mct-locally

scottbell commented:
> @scottbell I mean that we should document this behavior outside of the webpack, e2e, and karma files and in some place for others to reference.
> 
> https://github.com/nasa/openmct#building-and-running-open-mct-locally

Ah, gotcha. Can we make a ticket to do this please?

akhenry commented:
> Note minimist is still being used by gendocs.js.

We haven't used gendocs for some time. Let's get rid of it please.

scottbell commented:
> > Note minimist is still being used by gendocs.js.
> 
> We haven't used gendocs for some time. Let's get rid of it please.

It looks like we don't directly depend on it, but I've removed `gendocs.sh` and `header.html` and `footer.html` for good measure.

unlikelyzero commented on package.json:
```diff
@@ -69,14 +67,14 @@
     "clean-test-lint": "npm run clean; npm install; npm run test; npm run lint",
-    "start": "node app.js",
+    "start": "npx webpack serve --config ./webpack.dev.js",
+    "start:coverage": "npx webpack serve --config ./webpack.coverage.js",
```
super clean

unlikelyzero commented:
@scottbell https://github.com/nasa/openmct/blob/bbb551f34c65b4ea46670352dbe9d05ab653d4c1/karma.conf.js#L49 here's another NODE_ENV definition.

scottbell commented:
> @scottbell
> 
> https://github.com/nasa/openmct/blob/bbb551f34c65b4ea46670352dbe9d05ab653d4c1/karma.conf.js#L49
> here's another NODE_ENV definition.

@unlikelyzero I this one is ok as we're explicitly flagging it in the `package.json`:
```
"test:debug": "cross-env NODE_ENV=debug karma start --no-single-run"
```

unlikelyzero commented:
So that one is just interpreted at run time and requires that development is run at build time?

scottbell commented:
> So that one is just interpreted at run time and requires that development is run at build time?

I'm not sure I follow 😀 The environment flag is added by the script to start the karma unit tests in debug mode. Karma sees the flag and starts Chrome with a head.

unlikelyzero commented on e2e/playwright-local.config.js:
```diff
@@ -15,7 +15,7 @@ const config = {
         env: {
             NODE_ENV: 'test'
```
@ozyx is this necessary if we're running with a new `npm run start:coverage` ?

ozyx replied:
@scottbell @unlikelyzero :
This should not be necessary anymore since we were just using the envvar to control when webpack HMR launched (also coverage). We should also remove the `cross-env` logic from the other configs

unlikelyzero replied:
@scott where is HMR configured?

scottbell replied:
@unlikelyzero in [this block](https://github.com/nasa/openmct/pull/5797/files#diff-52cf7951fec7ee5c8db1888ab977bc94c938511c211beecca96a1ebb64a4dafaR38). Per the [webpack docs](https://webpack.js.org/configuration/dev-server/#devserverhot):

> Since webpack-dev-server v4, HMR is enabled by default. It automatically applies [webpack.HotModuleReplacementPlugin](https://webpack.js.org/plugins/hot-module-replacement-plugin/) which is required to enable HMR. So you don't have to add this plugin to your webpack.config.js when hot is set to true in config or via the CLI option --hot. See the [HMR concepts page](https://webpack.js.org/concepts/hot-module-replacement/) for more information.

ozyx commented on e2e/playwright-performance.config.js:
```diff
@@ -11,7 +11,7 @@ const config = {
     workers: 1, //Only run in serial with 1 worker
     webServer: {
-        command: 'cross-env NODE_ENV=test npm run start',
+        command: 'cross-env NODE_ENV=test npm run start:coverage',
```
```suggestion
        command: 'npm run start:coverage',
```

ozyx commented on e2e/playwright-visual.config.js:
```diff
@@ -10,7 +10,7 @@ const config = {
     workers: 1, //Lower stress on Circle CI Agent for Visual tests https://github.com/percy/cli/discussions/1067
     webServer: {
-        command: 'cross-env NODE_ENV=test npm run start',
+        command: 'cross-env NODE_ENV=test npm run start:coverage',
```
```suggestion
        command: 'npm run start:coverage',
```

ozyx commented on webpack.common.js:
```diff
@@ -1,5 +1,13 @@
+This is the OpenMCT common webpack file. It is imported by the other three webpack configurations:
+ - webpack.prod.js - the production configuration for OpenMCT (default)
+ - webpack.dev.js - the development configuration for OpenMCT
+ - webpack.coverage.js - imports webpack.dev.js and adds code coverag
```
e

ozyx commented on webpack.common.js:
```diff
@@ -1,5 +1,13 @@
-/* global __dirname */
+/* global __dirname module */
```
awesome

ozyx commented on e2e/playwright-ci.config.js:
```diff
@@ -14,7 +14,7 @@ const config = {
     timeout: 60 * 1000,
     webServer: {
-        command: 'cross-env NODE_ENV=test npm run start',
+        command: 'cross-env NODE_ENV=test npm run start:coverage',
```
```suggestion
        command: 'npm run start:coverage',
```

ozyx commented on e2e/playwright-local.config.js:
```diff
@@ -15,7 +15,7 @@ const config = {
         env: {
```
This entire `env:` block can be removed since it is no longer needed to enable coverage / disable webpack HMR

ozyx reviewed (changes requested):
Awesome stuff! Just have a couple suggestions for config cleanup

unlikelyzero commented:
@scottbell ok, after some more tinkering, I was able to remove almost all of our runtime cli evaluation and clean up some of our scripting along the way. Since runtime env var evaluation is gone, nothing really needs to be documented.

unlikelyzero commented on package.json:
```diff
@@ -69,23 +66,23 @@
     "build:coverage": "webpack --config webpack.coverage.js",
     "build:watch": "webpack --config webpack.dev.js --watch",
     "info": "npx envinfo --system --browsers --npmPackages --binaries --languages --markdown",
-    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--max_old_space_size=4096\" karma start --single-run",
```
node memory limits increased to 4GB with node 14 so we no longer need to set this at runtime

unlikelyzero commented on package.json:
```diff
@@ -69,23 +66,23 @@
-    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--max_old_space_size=4096\" karma start --single-run",
-    "test:debug": "cross-env NODE_ENV=debug karma start --no-single-run",
+    "test": "karma start",
+    "test:debug": "KARMA_DEBUG=true karma start",
```
just turned this into its own var since `NODE_ENV` isn't necessary

unlikelyzero reviewed (approved):
ok, I think i fixed everything related to runtime env var evaluation

scottbell commented:
@akhenry i think we're waiting for one last look over from you and this PR is ready to go

unlikelyzero commented on karma.conf.js:
```diff
@@ -70,6 +88,7 @@ module.exports = (config) => {
         },
         coverageIstanbulReporter: {
             fixWebpackSourcePaths: true,
+            skipFilesWithNoCoverage: true,
```
@akhenry this may fix some of the code coverage inconsistency that we're seeing in our reports

ozyx commented on e2e/playwright-visual.config.js:
```diff
@@ -4,13 +4,13 @@
 /** @type {import('@playwright/test').PlaywrightTestConfig<{ theme: string }>} */
 const config = {
-    retries: 1, // visual tests should never retry due to snapshot comparison errors. Leaving as a shim
+    retries: 1, // Visual tests should never retry due to snapshot comparison errors. Leaving as a shim
```
I think we can set this back to `0` now since the Percy timeout bug is resolved.

ozyx commented on package.json:
```diff
@@ -52,6 +51,7 @@
     "moment-timezone": "0.5.37",
     "nyc": "15.1.0",
     "painterro": "1.2.78",
+    "playwright-core": "1.26.1",
```
do we need this?

ozyx replied:
which version of `npm` are you using? I removed it and did a clean + install and didn't see any warnings

scottbell replied:
It happens post webpack compile. Note
`npm WARN @percy/playwright@1.0.4 requires a peer of playwright-core@>=1 but none is installed. You must install peer dependencies yourself.`
below:

```
$ npm run clean

> openmct@2.1.1-SNAPSHOT clean /Users/scott/workspace/openmct
> rm -rf ./dist ./node_modules ./package-lock.json

✔ ~/workspace/openmct [5031-replace-appjs-with-webpack-dev-server|✚ 1] 
19:08 $ npm install
npm WARN deprecated codecov@3.8.3: https://about.codecov.io/blog/codecov-uploader-deprecation-plan/

> core-js@3.25.3 postinstall /Users/scott/workspace/openmct/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> @percy/core@1.10.3 postinstall /Users/scott/workspace/openmct/node_modules/@percy/core
> node ./post-install


> openmct@2.1.1-SNAPSHOT prepare /Users/scott/workspace/openmct
> npm run build:prod


> openmct@2.1.1-SNAPSHOT build:prod /Users/scott/workspace/openmct
> webpack --config webpack.prod.js

webpack compiled successfully
npm WARN @percy/playwright@1.0.4 requires a peer of playwright-core@>=1 but none is installed. You must install peer dependencies yourself.
npm WARN ajv-keywords@3.5.2 requires a peer of ajv@^6.9.1 but none is installed. You must install peer dependencies yourself.

added 771 packages from 1200 contributors and audited 774 packages in 54.989s

86 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

scottbell replied:
The `ajv` requirement is impossible to meet as far as I can tell.

unlikelyzero replied:
@scottbell this is installed separately in the `npx playwright@version install` step

ozyx replied:
I only see this warning with node 14 and since that's the lower bound of what we support, I suppose let's keep this?

scottbell replied:
Agreed

ozyx replied:
@scottbell  I think we can resolve the `ajv` warning as well by adding

`"ajv": "6.9.1",`

unlikelyzero replied:
@scottbell do you mind removing the core pin? It'll be installed from other sources

scottbell replied:
Ah I remember now:
```
$ npm run clean && npm install

> openmct@2.1.1-SNAPSHOT clean /Users/scott/workspace/openmct
> rm -rf ./dist ./node_modules ./package-lock.json

npm WARN deprecated codecov@3.8.3: https://about.codecov.io/blog/codecov-uploader-deprecation-plan/

> core-js@3.25.3 postinstall /Users/scott/workspace/openmct/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> @percy/core@1.10.3 postinstall /Users/scott/workspace/openmct/node_modules/@percy/core
> node ./post-install


> openmct@2.1.1-SNAPSHOT prepare /Users/scott/workspace/openmct
> npm run build:prod


> openmct@2.1.1-SNAPSHOT build:prod /Users/scott/workspace/openmct
> webpack --config webpack.prod.js

webpack compiled successfully
added 780 packages from 1200 contributors and audited 783 packages in 52.341s

86 packages are looking for funding
  run `npm fund` for details

found 1 moderate severity vulnerability
  run `npm audit fix` to fix them, or `npm audit` for details
```

it gives a "moderate severity vulnerability" if we add `ajv` `6.9.1`

ozyx replied:
resolution is to keep the `ajv` warning because the required version produces an audit warning, and upgrading to resolve the audit warning breaks webpack.

scottbell replied:
> @scottbell do you mind removing the playwrightcore pin? It'll be installed from other sources

If we do this, we get a warning from `npm` that we're not installing it:
```
$ npm run clean && npm install

> openmct@2.1.1-SNAPSHOT clean /Users/scott/workspace/openmct
> rm -rf ./dist ./node_modules ./package-lock.json

npm WARN deprecated codecov@3.8.3: https://about.codecov.io/blog/codecov-uploader-deprecation-plan/

> core-js@3.25.3 postinstall /Users/scott/workspace/openmct/node_modules/core-js
> node -e "try{require('./postinstall')}catch(e){}"


> @percy/core@1.10.3 postinstall /Users/scott/workspace/openmct/node_modules/@percy/core
> node ./post-install


> openmct@2.1.1-SNAPSHOT prepare /Users/scott/workspace/openmct
> npm run build:prod


> openmct@2.1.1-SNAPSHOT build:prod /Users/scott/workspace/openmct
> webpack --config webpack.prod.js

webpack compiled successfully
npm WARN @percy/playwright@1.0.4 requires a peer of playwright-core@>=1 but none is installed. You must install peer dependencies yourself.
npm WARN ajv-keywords@3.5.2 requires a peer of ajv@^6.9.1 but none is installed. You must install peer dependencies yourself.

added 771 packages from 1200 contributors and audited 774 packages in 55.911s

86 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
```

ozyx reviewed (changes requested):
Looking really good. Have one question and small suggestion for config change.

ozyx reviewed (approved):
LGTM

shefalijoshi commented:
Verified that there are no more cross-env and app.js in the openmct package and webpack-dev-server is being used and included in package.json
